### ISSUE TITLE
Cop idea: Avoid tap

### ISSUE THREAD

**Issue by ydakuka** [rubocop/rubocop#12342]
## Description

The method `tap` can be helpful for debugging purposes but should not be left in production code.

https://rubystyle.guide/#avoid-tap

## Describe the solution you'd like

```ruby
# bad
Config.new(hash, path).tap do |config|
  config.check if check
end

# good
config = Config.new(hash, path)
config.check if check
config
```

**Comment by fatkodima** [comment]
> The method tap can be helpful for debugging purposes but should not be left in production code.

It is not only used for debugging and is perfectly fine to left in the production code. Using `tap` is to make the code 1 line less. It is heavily used in rails, for example. But I agree, without `tap`, most of the time it looks better.

**Comment by koic** [comment]
This looks at cases where `tap` is used to reduce verbose writing. I'm uncertain if opinions are uniform enough to enforce this in the style guide.

**Comment by seandunn** [comment]
In my opinion using `tap` for anything other than debugging is misuse.

Whilst it's "cute" to use it as an alternative to assignment you're wasting memory and CPU cycles for good reason _and_ adding noise to your code.

It doesn't really make code less verbose, but it does cause you to assign a extra block and throw it way on every call.

Again, IMHO, that's not elegant it's just naive.
